openml/runs/trace.py

```diff
@@ -26,56 +30,348 @@ def get_selected_iteration(self, fold, repeat):
         """
-        Logic to deserialize the trace from the filesystem
+        Logic to deserialize the trace from the filesystem.
 
         Parameters
         ----------
         file_path: str
-            File path where the trace is stored
+            File path where the trace arff is stored.
 
         Returns
         ----------
-        trace: dict
-            a dict in the liac-arff style that contains trace information
+        OpenMLRunTrace
         """
         if not os.path.isfile(file_path):
             raise ValueError('Trace file doesn\'t exist')
 
         with open(file_path, 'r') as fp:
-            trace = arff.load(fp)
+            trace_arff = arff.load(fp)
 
-        # TODO probably we want to integrate the trace object with the run object, rather than the current
-        # situation (which stores the arff)
-        for trace_idx in range(len(trace['data'])):
-            # iterate over first three entrees of a trace row (fold, repeat, trace_iteration) these should be int
+        for trace_idx in range(len(trace_arff['data'])):
+            # iterate over first three entrees of a trace row
+            # (fold, repeat, trace_iteration) these should be int
             for line_idx in range(3):
-                value = trace['data'][trace_idx][line_idx]
-                trace['data'][trace_idx][line_idx] = int(trace['data'][trace_idx][line_idx])
-        return trace
+                trace_arff['data'][trace_idx][line_idx] = int(
+                    trace_arff['data'][trace_idx][line_idx]
+                )
+
+        return cls.trace_from_arff(trace_arff)
+
+    def to_filesystem(self, file_path):
+        """Serialize the trace object to the filesystem.
+
+        Serialize the trace object as an arff.
+
+        Parameters
+        ----------
+        file_path: str
+            File path where the trace arff will be stored.
+        """
+
+        trace_arff = arff.dumps(self.trace_to_arff())
+        with open(os.path.join(file_path, 'trace.arff'), 'w') as f:
+            f.write(trace_arff)
+
+    def trace_to_arff(self):
+        """Generate the arff dictionary for uploading predictions to the server.
+
+        Uses the trace object to generate an arff dictionary representation.
+
+        Returns
+        -------
+        arff_dict : dict
+            Dictionary representation of the ARFF file that will be uploaded.
+            Contains information about the optimization trace.
+        """
+        if self.trace_iterations is None:
+            raise ValueError("trace_iterations missing from the trace object")
+
+        # attributes that will be in trace arff
+        trace_attributes = [
+            ('repeat', 'NUMERIC'),
+            ('fold', 'NUMERIC'),
+            ('iteration', 'NUMERIC'),
+            ('evaluation', 'NUMERIC'),
+            ('selected', ['true', 'false']),
+        ]
+        trace_attributes.extend([
+            ('parameter_' + parameter, 'STRING') for parameter in
```

Comment:
please use the constant (that might have become a static class variable by now) here